src/emitter/batch_emitter.rs

```diff
@@ -0,0 +1,318 @@
+
+use crate::emitter::Emitter;
+use crate::error::Error;
+use crate::event_batch::EventBatch;
+use crate::event_store::DEFAULT_EVENT_STORE_CAPACITY;
+use crate::event_store::{EventStore, InMemoryEventStore};
+use crate::http_client::ReqwestClient;
+use crate::payload::PayloadBuilder;
+use crate::HttpClient;
+
+/// An implementation of the [Emitter] trait that sends batched events to the Snowplow Collector.
+pub struct BatchEmitter {
+    /// The URL of your Snowplow [Collector](https://docs.snowplow.io/docs/pipeline-components-and-applications/stream-collector/)
+    collector_url: String,
+    /// A [HttpClient](crate::HttpClient) implementation to send events to the Snowplow Collector
+    http_client: Box<dyn HttpClient + Send + Sync>,
+    /// An [EventStore](crate::EventStore) implementation, used to queue events
+    event_store: Arc<Mutex<dyn EventStore + Send + Sync>>,
+    /// The thread running the tokio runtime
+    executor_handle: Option<std::thread::JoinHandle<()>>,
+    /// The transmitter to send an [EmitterMessage] to the [Emitter] thread
+    tx: tokio::sync::mpsc::Sender<EmitterMessage>,
+}
+
+/// Possible messages to send to the Emitter, sent via the [Emitter] transmitter
+#[derive(Debug)]
+pub enum EmitterMessage {
+    /// Sends a batch of events
+    Send(EventBatch),
+    /// Shuts down the [Emitter]
+    /// This will also attempt to send all events currently in the [EventStore]
+    Close,
+}
+
+/// A builder for the [BatchEmitter] struct
+pub struct BatchEmitterBuilder {
+    collector_url: Option<String>,
+    event_store: Arc<Mutex<dyn EventStore + Send + Sync>>,
+}
+
+impl BatchEmitterBuilder {
+    pub fn default() -> Self {
+        Self {
+            collector_url: None,
+            event_store: Arc::new(Mutex::new(InMemoryEventStore::default())),
+        }
+    }
+
+    /// Set the URL of your Snowplow [Collector](https://docs.snowplow.io/docs/pipeline-components-and-applications/stream-collector/)
+    pub fn collector_url(mut self, collector_url: &str) -> Self {
+        self.collector_url = Some(collector_url.to_string());
+        self
+    }
+
+    /// Set the [EventStore] implementation  
+    pub fn event_store(mut self, event_store: impl EventStore + Send + Sync + 'static) -> Self {
+        self.event_store = Arc::new(Mutex::new(event_store));
+        self
+    }
+
+    /// Build the [BatchEmitter]
+    pub fn build(self) -> Result<BatchEmitter, Error> {
+        match self.collector_url {
+            Some(collector_url) => {
+                let event_store_capacity = match self.event_store.lock() {
+                    Ok(event_store) => event_store.capacity(),
+                    Err(e) => {
+                        return Err(Error::EventStoreError(
+                            format!("Failed to lock event store: {}", e).to_string(),
+                        ))
+                    }
+                };
+
+                let (tx, mut rx) = tokio::sync::mpsc::channel(event_store_capacity);
```

Comment:
This initialization logic seems to be repeated here and in the `BatchEmitter::new()` method – can we call `new()` from here to avoid duplicating it?